### ISSUE TITLE
One more 4byte emoji issue

### ISSUE THREAD

**Issue by hachi8833** [dlclark/regexp2#6]
Hello dlclark, 

I found one more issue, which is relevant to #5 .

I checked this with the brand-new Go 1.8, but I guess the version of the Go does not affect the issue. 

## 1. Sample that encounter panic

Condition: 

* regex pattern contains "[猟](http://www.fileformat.info/info/unicode/char/731f/index.htm)"(U+731F), followed by a Japanese character like Hiragana
* target string contains 4byte emojis, followed by the same Japanese characters as above (Hiragana in the case)
* the 4byte emojis in the target is *not* adjacent to the same Japanese characters behind it

```go
package main

import (
	"github.com/dlclark/regexp2"
	"github.com/k0kubun/pp"
)

func main() {
	str := "猟な" // factor 1: the kanji + the Hiragana

	re := regexp2.MustCompile(str, 0)
	result, _ := re.ReplaceFunc(
		"🍺な" + // factor 2: 4byte emoji + the same Hiragana as above
                "なあ🍺な", // factor 3: the same Hiragana does not surround the 4byte emoji; if you remove "あ" from this, it works fine
		func(m regexp2.Match) string {
			return "࿗" + "1" + "࿘" + string(m.Capture.Runes()) + "࿌"
		}, -1, -1)

	pp.Println(result)
}
```

## 2. Sample that works fine

```go
package main

import (
	"github.com/dlclark/regexp2"
	"github.com/k0kubun/pp"
)

func main() {
	str := "猟な" // works fine with the kanji + trailing Japanese char ("な" in the case)

	re := regexp2.MustCompile(str, 0)
	result, _ := re.ReplaceFunc(
		"な📍な"+ // works fine if the same "な" surrounds the 4byte emoji
			"な✔️な"+
			"な😏な"+
			"な⚾️な"+
			"な📣な"+
			"な🍣な"+
			"な🍺な"+
			"な📍✔️😏⚾️📣🍣🍺な", func(m regexp2.Match) string {
			return "࿗" + "1" + "࿘" + string(m.Capture.Runes()) + "࿌"
		}, -1, -1)

	pp.Println(result)
}
```

Best regards, 🙇

**Comment by dlclark** [comment]
Thanks for finding this.  In my fix for #5 I had missed a duplicative section of the Boyer-Moore prefix Scan code.  

How are you finding these?  Do you know if there's an extended unicode test battery I can use to try and find these issues earlier?

**Comment by hachi8833** [comment]
Amazing! Thank you very much 🙇 I appreciate for your work.

Onigmo's test code might help you https://github.com/k-takata/Onigmo/blob/master/testpy.py
